contracts/CPKFactory.sol

```diff
@@ -1,54 +1,69 @@
 pragma solidity >=0.5.0 <0.7.0;
 
-import { Enum } from "@gnosis.pm/safe-contracts/contracts/common/Enum.sol";
-import { Proxy } from "@gnosis.pm/safe-contracts/contracts/proxies/Proxy.sol";
+import { GnosisSafeProxy } from "@gnosis.pm/safe-contracts/contracts/proxies/GnosisSafeProxy.sol";
 import { GnosisSafe } from "@gnosis.pm/safe-contracts/contracts/GnosisSafe.sol";
 
+import { ProxyImplSetter } from "./ProxyImplSetter.sol";
+
 contract CPKFactory {
-    event ProxyCreation(Proxy proxy);
+    //keccak256(
+    //    "EIP712Domain(address verifyingContract)"
+    //);
+    bytes32 private constant DOMAIN_SEPARATOR_TYPEHASH = 0x035aff83d86937d35b32e04f0ddc6ff469290eef2f1b692d8a815c89404d4749;
+
+    //keccak256(
+    //    "SafeTx(address to,uint256 value,bytes data,uint8 operation,uint256 safeTxGas,uint256 baseGas,uint256 gasPrice,address gasToken,address refundReceiver,uint256 nonce)"
+    //);
+    bytes32 private constant SAFE_TX_TYPEHASH = 0xbb8310d486368db6bd6f849402fdd73ad53d316b5a4b2644ad6efe0f941286d8;
+
+    ProxyImplSetter public proxyImplSetter;
+
+    constructor() public {
+        proxyImplSetter = new ProxyImplSetter();
+    }
+
+    event ProxyCreation(GnosisSafeProxy proxy);
 
     function proxyCreationCode() external pure returns (bytes memory) {
-        return type(Proxy).creationCode;
+        return type(GnosisSafeProxy).creationCode;
     }
 
     function createProxyAndExecTransaction(
-        address masterCopy,
+        address owner,
+        address safeVersion,
         uint256 saltNonce,
         address fallbackHandler,
-        address to,
-        uint256 value,
-        bytes calldata data,
-        Enum.Operation operation
+        bytes calldata execTxCalldata
     )
         external
+        payable
         returns (bool execTransactionSuccess)
     {
-        GnosisSafe proxy;
-        bytes memory deploymentData = abi.encodePacked(type(Proxy).creationCode, abi.encode(masterCopy));
-        bytes32 salt = keccak256(abi.encode(msg.sender, saltNonce));
-        // solium-disable-next-line security/no-inline-assembly
-        assembly {
-            proxy := create2(0x0, add(0x20, deploymentData), mload(deploymentData), salt)
+        bytes32 salt = keccak256(abi.encode(owner, saltNonce));
+
+        address payable proxy;
+        {
+            bytes memory deploymentData = abi.encodePacked(
```

Comment:
Theoretically we could use the existing proxy factory for the proxy creation, makes this a little easier to monitor by our services.